### ISSUE TITLE
LCCSCF_PIPELINE pipeline stop working when using an IP address to connect

### ISSUE THREAD

**Issue by snaami** [warmcat/libwebsockets#2575]
Hello,
I noticed some wired behavior when using `LCCSCF_PIPELINE` in order to benefit from pipelining:

-If using tls and info.address = info.origin = info.host it's work just fine only one client_hello in Wireshark
Note here I had to add `LCCSCF_SKIP_SERVER_CERT_HOSTNAME_CHECK` to not stop on hostname validation

`info.address = "46.105.127.147"; //libwebsockets.org ip`
`info.host = info.address;`
`info.origin = info.address;`

but if we try to use for example the IP and still want to send the right host header it stops working

`info.address = "46.105.127.147"; //libwebsockets.org ip`
`info.host = "libwebsockets.org";`
`info.origin = "libwebsockets.org";`

Here when using the `minimal-http-client-multi` I see 8 different TLS session (8 client_hello and key exchanging).

If this is not a bug, and it's the desired behavior, so my question is how I can use pipelining with an IP address and keep having the possibility to validate the host name and having the right host in HTTP headers ?

-if no TLS is used, I noticed that pipelining is working only if  `info.address = info.host = ip_address ` so again the host in the HTTP headers will not be correct.

In my case, I can't use the DNS service provided by libwebsockets, I need to do some other checks on DNS cache myself.

**Comment by lws-team** [comment]
> If this is not a bug, and it's the desired behavior, so my question is how I can use pipelining with an IP address and keep having the possibility to validate the host name and having the right host in HTTP headers ?

There is some confusion in this question as it is... the peer certificate contains the host NAME in the CN field that's used to check it.  `LCCSCF_SKIP_SERVER_CERT_HOSTNAME_CHECK` already makes your situation completely insecure... for example if you trust the Let's Encrypt CA, then if the server at the IP address just gives you someone else's LE certificate that is validly signed for "whatever.com" instead of your actual hostname, you will gladly and fully accept it as a valid cert for your IP address.  It basically breaks any validation action to do that.

You should not do any of that if you care about peer validation, instead, you should arrange that the hostname DNS resolves to the IP address you want.  For Linux, at its simplest you can just add an entry in `/etc/hosts` for the hostname.  If that is an attack, the client will not find a trusted cert for the hostname at the attacker's box and validation will fail (correctly).  If it's true, the real server is at that IP address, then it will find a trusted cert for the right hostname and succeed.

In other words you must give it the correct hostname as the address when setting up the connection for tls (and then, pipelining) to work properly.  Any IP guidance must be done at DNS.

> In my case, I can't use the DNS service provided by libwebsockets, I need to do some other checks on DNS cache myself.

By default, lws uses libc gethostbyname(), that can be configured to do DNSSEC depending on your host.  Lws also provides Async DNS option it does internally, it parses the udp DNS response itself.  What additional checks do you have in mind?

**Comment by snaami** [comment]
Hello thanks for the reply, and sorry if the question is confusing.
Actually, the DNS is done via c-ares, and it's not my main issue here. 
All I want to do is to connect via an actual ip address and still having pipeline for my connection, so what I did is configuring the connection info like this:
`info.address = "46.105.127.147"; //libwebsockets.org ip`
`info.host = "libwebsockets.org";`
`info.origin = "libwebsockets.org";`
as I already know the ip of my server but with that there was no connection  pipeline, the pipeline is working only if 
`info.host = info.address`
you are right, I get confused with the ` LCCSCF_SKIP_SERVER_CERT_HOSTNAME_CHECK` it has nothing to do with the problem. 
the problem is how I can set the host header to something other than the IP and still having pipeline.

**Comment by lws-team** [comment]
> All I want to do is to connect via an actual ip address and still having pipeline for my connection, 

As I described, you cannot do that and still have any meaningful tls protection.  So I think this question reflects some other confusion going on.

**Comment by snaami** [comment]
Yes  I see the problem now, I was a bit confused, I will try to avoid this situation for now I will giveup on connection pipeline until I figure out a better approach to redesign my application.  Thank you for the help, and I think we can close this issue.

**Comment by lws-team** [comment]
... connection via pipeline isn't your problem.  Your problem is you cannot validate who you are talking to when you coonect to an IP address rather than a hostname.  Anyway... closing.

**Comment by snaami** [comment]
Even if it's closed, I see : 
`	LWS_CALLBACK_OPENSSL_LOAD_EXTRA_CLIENT_VERIFY_CERTS	= 21,`
	/**< if configured for
	 * including OpenSSL support, this callback allows your user code
	 * to perform extra SSL_CTX_load_verify_locations() or similar
	 * calls to direct OpenSSL where to find certificates the client
	 * can use to confirm the remote server identity.  user is the
	 * OpenSSL SSL_CTX* */
can't I use this callback to perform some extra checks on hostname?

**Comment by lws-team** [comment]
It's the job of the tls library to judge if the thing you connected to should be trusted or not.  There's really no point thinking that you know better than it does what should be trusted.

The right way to do this is just tell lws the hostname as the address to connect to, lws will arrange that the tls library will insist to see that name in the server's CN field and check that a trusted CA signed the cert with the right name.  The DNS name and the certificate name are designed to be the same, for DNS lookup and for cert validation purposes.  The idea you should tell it an IP address at all seems to be where you are going off into the weeds... just don't do that.

**Comment by snaami** [comment]
That is my problem I can't use the hostname for connection, I need to keep track of all DNS entries myself and their TTL and also when to use IPv4 and when to use IPv6 to connect as I will have multiple servers but all with the same url, and it's the client job to balance the traffic between them so my DNS response will have multiple IPs each with different TTL and I need to use one of them and then keep track of its TTL etc .... there is much stuff to be done with the DNS response before trying to connect and by using the hostname in connection info I can't control all of that. Thank you so much for the help, it really helped me get better understanding of the lib and the situation I have.

**Comment by lws-team** [comment]
> That is my problem I can't use the hostname for connection,

You seem to want some kind of DNS load-balancer... if you must do it at the client, then I would get rid of c-ares and look at LWS_WITH_SYS_ASYNC_DNS.

Tell lws the hostname (as you must), let it do the DNS resolution and hook that

https://libwebsockets.org/git/libwebsockets/tree/lib/system/async-dns/async-dns-parse.c

it knows the TTL of the response it is interested in, it already maintains a local cache of results that is aware of and maintains the cache by individual TTLs.  I am not sure what else you want to depend on the TTL but that code in lws already knows it if you use Async DNS.  If it's useful, I don't mind adding a hook in the lws_system struct so you can wire it up without patching lws afterwards.

TTL is subject to manipulation by downstream DNS caches, I am not sure what meaning you are trying to overload out of it but you'll have to bear that in mind.

**Comment by snaami** [comment]
Thank you, I will take a look at LWS_WITH_SYS_ASYNC_DNS and see what I can do, maybe I can get rid of c-ares.
